LUCENE-10342: Add some logging to static initializers to warn users if unmapping or object size calculation does not work

uschindler opened the pull request:
Also give hints how to fix. This also removes the hard requirement to "jdk.unsupported" module. This now causes Luke (it's a good test to log the warning in the UI -> @mocobeta). 

See https://issues.apache.org/jira/browse/LUCENE-10342 for details.

I have not yet added a InfoStream implementation based on java.util.logging, but that's a good chance to do this.

uschindler committed:
LUCENE-10342: Add some logging to static initializers to warn users if unmapping or object size calculation does not work. Also give hints how to fix.

uschindler commented on lucene/test-framework/src/java/org/apache/lucene/tests/util/LuceneTestCase.java:
```diff
@@ -554,7 +554,7 @@ public boolean shouldCache(Query query) throws IOException {
         maxFailures = 1;
       } else {
-        Logger.getLogger(LuceneTestCase.class.getSimpleName())
+        Logger.getLogger(LuceneTestCase.class.getName())
```
I just modified this to look like everything else in Lucene.

uschindler replied:
This is the only place in test-framework that uses a logger. All other places use System.err. Should we not move to this and remove logging (also dependency to module).

uschindler replied:
I checked Git Blame, it slurped in 10 years ago, its completely obsolete. I will replace it and remove logging from test-framework.

We should either log everything or nothing, but not a mix.

uschindler commented:
@dweiss: The core modular test still succeeeds, although the jdk.unsupported module is not added with "requires", but "requires static". I have no idea why this is so. Could it be that Gradle adds jdk.unsupported by default when starting test runner? If I write a normal modular application it works as expected (see luke), but when testing jdk.unsupported is added automatically.

dweiss commented:
bq. Could it be that Gradle adds jdk.unsupported by default when starting test runner?

You can run gradle with debug option and see what the test worker command line is, I really don't know. Also, we use "--add-modules", "ALL-MODULE-PATH" - perhaps this loads it?

uschindler committed:
Add a JavaLoggingInfoStream for convenience (it's a copy from a private project that I will publish here)

uschindler commented:
> bq. Could it be that Gradle adds jdk.unsupported by default when starting test runner?
> 
> You can run gradle with debug option and see what the test worker command line is, I really don't know. Also, we use "--add-modules", "ALL-MODULE-PATH" - perhaps this loads it?

The interesting thing is: If I remove the requires directive completey, it fails. So it looks like the following is happening: the ALL-MODULE-PATH makes it visible to the module layer. Because it is visible there, the JDK adds it and the optional dependency gets resolved and satisfied. But I am just guessing.

uschindler commented:
I added a for-convenience InfoStream implementation using JUL logging.

uschindler committed:
Add a suppresswarning as explicitely don't make the logging dependency transitive in descriptor

dweiss commented:
bq. If I remove the requires directive completey, it fails. So it looks like the following is happening: the ALL-MODULE-PATH makes it visible to the module layer. Because it is visible there, the JDK adds it and the optional dependency gets resolved and satisfied. But I am just guessing.

Hmm... but isn't requires static going to include the module, if it's available in the module graph (even if it's not explicitly added)? If you remove it completely - it fails and this seems to be fine. If you add it as a requires static, it should be included because it is available in the pool of modules. I think it works as intended?

uschindler commented:
> bq. If I remove the requires directive completey, it fails. So it looks like the following is happening: the ALL-MODULE-PATH makes it visible to the module layer. Because it is visible there, the JDK adds it and the optional dependency gets resolved and satisfied. But I am just guessing.
> 
> Hmm... but isn't requires static going to include the module, if it's available in the module graph (even if it's not explicitly added)? If you remove it completely - it fails and this seems to be fine. If you add it as a requires static, it should be included because it is available in the pool of modules. I think it works as intended?

No exactly: The JDK always contains the module, but if nothing else is not using the module it won't be added to module graph. You can try this with a plain luke: When you start Luke with current patch it logs the warning - as expected because I did not add `--add-module jdk.unsupported` willingly (and I won't - we have a warning printed and Luke does not need MMAP):

![image](https://user-images.githubusercontent.com/1005388/147503086-bbd533bd-ce2e-42d1-ac7f-bd64e473d77d.png)

So it looks like the `--add-modules ALL` adds it to module graph explicitly.

dweiss commented:
The rules are described in java.lang.module package - it seems that "Step 1" they talk about when discussing requires static would resolve static modules in the presence of any automatic module... But I'm really not sure. The wording isn't clear to me.

https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/lang/module/package-summary.html

dweiss commented:
Yeah, something is pulling that module into the set of loaded modules. Interesting.

uschindler committed:
Require non-null component name

rmuir commented on lucene/core/src/java/org/apache/lucene/store/MMapDirectory.java:
```diff
@@ -343,6 +344,9 @@ private IOException convertMapFailedIOException(
       UNMAP_NOT_SUPPORTED_REASON = hack.toString();
+      Logger.getLogger(MMapDirectory.class.getName())
+          .warning(
+              "Use of MMapDirectory was disabled by default: ".concat(UNMAP_NOT_SUPPORTED_REASON));
```
I think this is confusing to say 'MMAPDirectory was disabled' from `MMapDirectory` class? This class still functions correctly, it just won't forcibly unmap(), kernel resources may be wasted because GC has not decided they were important, etc. Still probably just fine for a static index. 

It is `FSDirectory` class that instead chooses a different default for `FSDirectory.open` based on whether unmap() is 
 supported and the platform is 64-bits. So I would prefer to log "enabled/disabled" from that class?

rmuir replied:
also its ok to do here if we just change the messaging to be less confusing. i am a little afraid of adding any clinit to fsdirectory that looks at mmapdirectory since it is the superclass of it. but maybe you know if it is really safe :)

uschindler replied:
I was tinking about this, too last night.

I'd just remove the message prefix and just print `UNMAP_NOT_SUPPORTED_REASON`. If this is caused by one calling `FSDirectory.open()` the person known that you may need to tune the unmapping (the reason and error messages are given). The log statement will also print class name, so all is fine.

Adding it as clinit toFSDirectory is bad, if somebody uses another subclass. The message could also move to `FSDirectory.open()`, but then it should be one-time only, so it needs an `AtomicBoolean` or similar :-)

So let's just remove the prefix on the message! OK?

rmuir replied:
sounds good, thanks

uschindler replied:
Fixed.

mocobeta commented on lucene/core/src/java/org/apache/lucene/util/RamUsageEstimator.java:
```diff
@@ -153,6 +154,19 @@ private RamUsageEstimator() {}
+          log.warning(
+              "Lucene cannot correctly calculate object sizes on 64bit JVMs, unless the 'jdk.management' Java module "
+                  + "is readable [please add 'jdk.management' to modular application either by command line or its module descriptor]");
+        } else {
+          log.warning(
+              "Lucene cannot correctly calculate object sizes on 64bit JVMs that are not based on Hotspot.");
```
+1. I checked this works for me (on Eclipse OpenJ9 VM AdoptOpenJDK-11.0.11+9); and without the explicit requirement directive for "jdk.management", this else clause can't be reached since Module.canRead() always returns false.

uschindler replied:
Thanks, in a followup issue I will try to make this work with OpenJ9. Looks like they have another MX Bean to get the JVM properties.

mocobeta replied:
> I will try to make this work with OpenJ9.

sorry, if my explanation was is not good.
Indeed this works for me with OpenJ9. I just found that in order to log this message on OpenJ9 "requires static jdk.management" is necessary and we already have that, so there is no problem.

uschindler replied:
Cool. So it detects compressed pointers correctly? I wanted to check this but it was too late to download J9.

uschindler replied:
I will do my own checks with J9 later, it is unrelated to this issue. I have the feeling we are talking about different things: The code as written here works (thanks for clarifying). With my comment I was asking you, if you have a quick info about J9 and support for a HotspotMXBean that can be used to detect reference size. If not, I will check for an alternative in a separate issue.

I will change the message a bit to say "Hotspot or a compatible implementation".

mocobeta replied:
What I see here is, OpenJ9 VM AdoptOpenJDK-11.0.11+9 (with off-the-shelf setup) throws NoSuchMethodException at this line:
https://github.com/apache/lucene/blob/e89c9fcca87b8febdecc581a1cd77ba479339176/lucene/core/src/java/org/apache/lucene/util/RamUsageEstimator.java#L136

Then the logger correctly outputs the warning message with this PR.
```
$ java --module-path build/libs -m modtest/modtest.Main
12月 29, 2021 11:57:30 午前 modtest.Main main
警告: Lucene cannot correctly calculate object sizes on 64bit JVMs that are not based on Hotspot.
java.lang.NoSuchMethodException: com.sun.management.HotSpotDiagnosticMXBean.getVMOption(java.lang.String)
        at java.base/java.lang.Class.newNoSuchMethodException(Class.java:635)

$ java --version
openjdk 11.0.11 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
Eclipse OpenJ9 VM AdoptOpenJDK-11.0.11+9 (build openj9-0.26.0, JRE 11 Linux amd64-64-Bit Compressed References 20210421_975 (JIT enabled, AOT enabled)
OpenJ9   - b4cc246d9
OMR      - 162e6f729
JCL      - 7796c80419 based on jdk-11.0.11+9)
```



I don't have much information about J9. I just have run applications on it once or twice for testing or making compatibility matrices.

uschindler replied:
Haha, the stub is empty interface: https://github.com/eclipse-openj9/openj9/blob/master/jcl/src/jdk.management/share/classes/com/sun/management/HotSpotDiagnosticMXBean.java

uschindler replied:
Looks like we can use this one: https://github.com/eclipse-openj9/openj9/blob/master/jcl/src/jdk.management/share/classes/openj9/lang/management/OpenJ9DiagnosticsMXBean.java

Will be a separate PR later. Thanks for trying.

uschindler replied:
Oh this one has no VM options, looks like 2012, impossible to get reference size.

dweiss commented:
> The core modular test still succeeeds, although the jdk.unsupported module is not added with "requires", but "requires static"

It's gradle test launcher. I'm not sure what it does or why (or where) but you can see the root set of modules the process starts with and it's, ahem, large.

```
gradlew -p lucene/core.tests "-Ptests.jvmargs=--show-module-resolution" test
```

for me it prints the following roots:
```
root java.sql jrt:/java.sql
root java.rmi jrt:/java.rmi
root jdk.management.jfr jrt:/jdk.management.jfr
root jdk.jdi jrt:/jdk.jdi
root org.apache.lucene.core.tests file:///C:/Work/apache/lucene/lucene/core.tests/build/classes/java/test/
root java.transaction.xa jrt:/java.transaction.xa
root java.xml.crypto jrt:/java.xml.crypto
root java.logging jrt:/java.logging
root java.xml jrt:/java.xml
root jdk.xml.dom jrt:/jdk.xml.dom
root jdk.jfr jrt:/jdk.jfr
root org.apache.lucene.core file:///C:/Work/apache/lucene/lucene/core/build/libs/lucene-core-10.0.0-SNAPSHOT.jar
root java.datatransfer jrt:/java.datatransfer
root jdk.httpserver jrt:/jdk.httpserver
root jdk.net jrt:/jdk.net
root java.desktop jrt:/java.desktop
root java.naming jrt:/java.naming
root java.prefs jrt:/java.prefs
root java.net.http jrt:/java.net.http
root jdk.compiler jrt:/jdk.compiler
root java.security.sasl jrt:/java.security.sasl
root jdk.jconsole jrt:/jdk.jconsole
root jdk.attach jrt:/jdk.attach
root java.base jrt:/java.base
root jdk.javadoc jrt:/jdk.javadoc
root junit file:///C:/Users/dweiss/.gradle/caches/modules-2/files-2.1/junit/junit/4.13.1/cdd00374f1fee76b11e2a9d127405aa3f6be5b6a/junit-4.13.1.jar automatic
root jdk.jshell jrt:/jdk.jshell
root java.sql.rowset jrt:/java.sql.rowset
root jdk.sctp jrt:/jdk.sctp
root java.management jrt:/java.management
root jdk.jsobject jrt:/jdk.jsobject
root jdk.unsupported jrt:/jdk.unsupported
root java.smartcardio jrt:/java.smartcardio
root jdk.scripting.nashorn jrt:/jdk.scripting.nashorn
root java.instrument jrt:/java.instrument
root java.security.jgss jrt:/java.security.jgss
root jdk.management jrt:/jdk.management
root jdk.security.auth jrt:/jdk.security.auth
root jdk.nio.mapmode jrt:/jdk.nio.mapmode
root java.compiler jrt:/java.compiler
root java.scripting jrt:/java.scripting
root jdk.dynalink jrt:/jdk.dynalink
root jdk.unsupported.desktop jrt:/jdk.unsupported.desktop
root jdk.accessibility jrt:/jdk.accessibility
root org.hamcrest file:///C:/Users/dweiss/.gradle/caches/modules-2/files-2.1/org.hamcrest/hamcrest/2.2/1820c0968dba3a11a1b30669bb1f01978a91dedc/hamcrest-2.2.jar automatic
root jdk.jartool jrt:/jdk.jartool
root java.management.rmi jrt:/java.management.rmi
root jdk.security.jgss jrt:/jdk.security.jgss
```

I don't know why these other roots are there - couldn't figure it out. A stand-alone check I did was working correctly with an automatic module on module path (junit), as well as with --add-modules ALL-MODULE-PATH.

dweiss commented:
Sorry - update - I know why. Java itself places all these modules in the root set when you're starting a class from classpath mode. So the test launcher (started as a classpath-main class) will bring in all these other modules. Confirmed by running a stand-alone simple example. 

```
# limited root set
java --module-path lib;mod --show-module-resolution --add-modules ALL-MODULE-PATH --module mod/test.Test
# full classpath-mode root set.
java --module-path lib;mod --show-module-resolution --add-modules ALL-MODULE-PATH --classpath mod test.Test
```

uschindler commented:
> So the test launcher (started as a classpath-main class) will bring in all these other modules. Confirmed by running a stand-alone simple example.

That explains. So we still need to work on running our tests as modules, too (by patching in the tests into the module like Maven does). Or is this a Gradle test runner limitation?

uschindler committed:
Improve test

dweiss commented:
> So we still need to work on running our tests as modules, too (by patching in the tests into the module like Maven does)

I don't think it's just about patching - it's also about a shadow module-info that includes additional modules (or additional arguments that do it from command-line). I lost the spirit there a bit - it's clumsy. That PR I posted an e-mail about contains modular tests, but they are placed in separate projects. By default src/test runs from classpath.

uschindler committed:
Just print the reason to logger

mocobeta commented:
> This now causes Luke (it's a good test to log the warning in the UI -> @mocobeta).

Thanks. I've been always wondering if integrating a GUI utility is a really good thing for the library project (it's already huge and having an application as a part of it adds additional consideration in terms of packaging). Luke certainly could have remained a separate project - but I think it would be great if it is somewhat helpful for internal tests/experiments.

uschindler committed:
Change message a bit to say "compatible".

uschindler committed:
Add the optional modules to test runners, so we make sure that Lucene is fully tested. At runtime, the dependencies can be left out, but for testing we want to make sure that they are available

uschindler commented:
@dweiss I added an explicit addition of both modules to test runner, so we are safe in future.

mocobeta commented:
> That PR I posted an e-mail about contains modular tests, but they are placed in separate projects. By default src/test runs from classpath.

White box tests have to remain in the classpath mode anyway, I think? Once we define which packages should be exported to outside modules and which ones are not (currently all existing packages are exported), it could be more clear to everyone what tests should reside in the modular tests...?

dweiss commented:
bq. White box tests have to remain in the classpath mode anyway, I think?

They don't have to be - you can "overlay" test classes over main or the other way around. What you can't do is "merge" two module descriptors - and you need additional opens/ module dependencies for running tests (think junit, randomized testing, etc.). It's very inconvenient to organize this as a pattern for "any" project, without knowing the internals.

dweiss commented:
a static require will fail much earlier if these modules are not available - it'll fail during compilation because at compile-time requires static is a hard requirement. So if it compiles, it'll likely pass tests without this explicit clause:
> "--add-modules", "jdk.unsupported,jdk.management"

dweiss commented:
That file you added it to has been severely rewritten on the PR branch - it'll probably cause a conflict.

dweiss commented:
This PR: https://github.com/apache/lucene/pull/571

mocobeta commented:
> They don't have to be - you can "overlay" test classes over main or the other way around. What you can't do is "merge" two module descriptors - and you need additional opens/ module dependencies for running tests

If my understanding is correct I was thinking similar things (opening packages in a module to its test module to allow reflective accesses for tests), but I'm a bit afraid that it could make the purpose of the modular tests unclear - it should be kept as black-box tests for APIs to outside modules to me. But this is a different concern here and should be discussed in another issue.

dweiss commented:
Both types of tests make sense and I don't think you can make them from the same codebase - these are effectively slightly different tests. There are three different "modes", effectively: classpath, modular (with module patching), full modules. We currently have 1 and 3.

uschindler commented:
> > They don't have to be - you can "overlay" test classes over main or the other way around. What you can't do is "merge" two module descriptors - and you need additional opens/ module dependencies for running tests
> 
> If my understanding is correct I was thinking similar things (opening packages in a module to its test module to allow reflective accesses for tests), but I'm a bit afraid that it could make the purpose of the modular tests unclear - it should be kept as black-box tests for APIs to outside modules to me. But this is a different concern here and should be discussed in another issue.

Actually the "patch-module" command line option allows to "patch" a module for testing (it was invented just for that). It allows to pass another JAR file or a file system directory that "overlays" the original module. So that means you can patch in all test classes into lucene.core, so they can run as if they were in the core module. This is how Maven manages the testing of modules (actually Maven people invented that in a meeting with Mark Reinhold, Alan Bateman and we discussed it including me on a OpenJDK committers meeting a few years ago).

As Dawid said, the problem is that test may need to patch module-info.java, too. I think this should work, too, but you have to have a separate module.info.java in the tests that gets patched over the existing one. I have not tested this. My problem is code duplication, but maybe we can use some ASM trick to add the "requires" directives or we do this on test startup using a config file (you can still change the module layer at runtime).

mocobeta commented:
I have seen (once or twice) looked `--patch-module` option and wondered what's the use-case... thanks for the detailed explanation.

uschindler commented:
> That file you added it to has been severely rewritten on the PR branch - it'll probably cause a conflict.

As a special service I will merge it for you after this PR gots merged. 😄

uschindler commented:
> > That file you added it to has been severely rewritten on the PR branch - it'll probably cause a conflict.
> 
> As a special service I will merge it for you after this PR gots merged. 😄

Alternatively I can add the option to test-defaults.gradle like in my other PR that uses the incubator modules. I just felt that it fits better in Module system part of build system.

What do you think?

dweiss commented:
> As a special service I will merge it for you after this PR gots merged.

LOL. :) I'm just saying I've done major changes to that file, that's all. Play a bit with that branch though - it shows non-trivial problems that are currently there (projects excluded from module-path treatment of dependencies). You'll find it interesting, I think.

uschindler committed:
Use a separate forbiddenapis file for restriction of logging:
- Luke is excluded, also not applied to test
- Cleanup the facet module and remove useless logging in production code
- Remove logging from spatial-extras code, but leave in tests (howto remove requires?)

uschindler commented:
Hi @rmuir: Please verify my latest changes about forbiddenapis: ed19120

@dweiss How to add "java.logging" module only to tests of spatial-extras without adding it to production module-info.java? Tests dont compile if you remove the now useless logging module. You can review here: ed19120

dweiss commented:
Let me take a look.

uschindler committed:
Remove dead code

dweiss commented:
I commented out java.logging in the descriptor and it works for me. Tests use classpath - they have full access to java logging and anything else. I'm not sure I follow where the problem is?

uschindler commented:
> I commented out java.logging in the descriptor and it works for me. Tests use classpath - they have full access to java logging and anything else. I'm not sure I follow where the problem is?

Do "gradlew clean"; it fails to compile the tests.

dweiss commented:
![image](https://user-images.githubusercontent.com/199470/147589774-f09c32f1-7000-44cd-85e5-4d4bd60fd986.png)

uschindler committed:
Cleanup test-framework

uschindler committed:
Remove logging from spatial-extras module-info.java, was a leftover import

uschindler commented:
Was a leftover import before I did "gradlew tidy". Now it compiles.

I also cleaned up the test framework, no logging anymore. Running tests again.

uschindler committed:
Make ECJ happy

uschindler commented:
I am fine+done with everything, only changes entry is missing.

uschindler committed:
Make ecj happy #2

uschindler committed:
Move "Add the optional modules to test runners, so we make sure that Lucene is fully tested" to defaults-tests.gradle where it better belongs (modules.gradle is generic code only, no custom Lucene specific stuff)

uschindler committed:
Add CHANGES.txt and some info in migration guide

rmuir reviewed (approved):
I like it: it should help users with the module system, without bringing crazy logging everywhere. Generally, as an API, we should throw exceptions, but for MMapDirectory/RAMUsageEstimator, it makes sense for them to not "fail hard" and instead just degrade a bit, logging a warning.

mocobeta commented on lucene/core/src/java/org/apache/lucene/util/JavaLoggingInfoStream.java:
```diff
@@ -0,0 +1,98 @@
+ *
+ * @lucene.internal
+ */
+// We export class 'java.util.logging.Level', but we won't don't want a transitive dependency in
```
I think this double negative is a typo...?
```suggestion
// We export class 'java.util.logging.Level', but we won't want a transitive dependency in
```

uschindler replied:
I fixed it, thanks for the "github commit suggestion" (never seen that before, everyday something new).

mocobeta commented:
I confirmed this log via luke running on OpenJ9 11.

```
17:32:51 [WARNING] org.apache.lucene.store.MMapDirectory: Unmapping is not supported, because Lucene cannot read 'jdk.unsupported' module [please add 'jdk.unsupported' to modular application either by command line or its module descriptor]
17:32:51 [WARNING] org.apache.lucene.util.RamUsageEstimator: Lucene cannot correctly calculate object sizes on 64bit JVMs that are not based on Hotspot or a compatible implementation.
17:32:52 [INFO] org.apache.lucene.luke.models.util.IndexUtils: IndexReaders (1 leaf readers) successfully opened. Index path=/home/moco/tmp/20ng
```

uschindler committed:
Update lucene/core/src/java/org/apache/lucene/util/JavaLoggingInfoStream.java

Co-authored-by: Tomoko Uchida <tomoko.uchida.1111@gmail.com>

uschindler commented:
> I confirmed this log via luke running on OpenJ9 11.
> 
> ```
> 17:32:51 [WARNING] org.apache.lucene.store.MMapDirectory: Unmapping is not supported, because Lucene cannot read 'jdk.unsupported' module [please add 'jdk.unsupported' to modular application either by command line or its module descriptor]
> 17:32:51 [WARNING] org.apache.lucene.util.RamUsageEstimator: Lucene cannot correctly calculate object sizes on 64bit JVMs that are not based on Hotspot or a compatible implementation.
> 17:32:52 [INFO] org.apache.lucene.luke.models.util.IndexUtils: IndexReaders (1 leaf readers) successfully opened. Index path=/home/moco/tmp/20ng
> ```
> 
> (It seems OpenJ9 resolves `com.sun.management.HotSpotDiagnosticMXBean` but the implementation actually don't have `getVMOption` method. Unfortunately, I don't have much information about it. [#572 (comment)](https://github.com/apache/lucene/pull/572#discussion_r776136859))

OK, thanks. I will check the compatibility with J9 later. It may also be that it has a VMOption, but the name is different (I have that in mind). So we can also try 2 different options for same.

This is unrelated to this issue, so we can fix it as a separate issue to bring more J9 compatibility. I did this last time around 2012 and the support for the bean was.... bad!
